// ngx-simple-modal/dist/simple-modal/simple-modal-holder.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var rxjs_1 = require("rxjs");
var simple_modal_options_1 = require("./simple-modal-options");
var simple_modal_wrapper_component_1 = require("./simple-modal-wrapper.component");
var SimpleModalHolderComponent = (function () {
    function SimpleModalHolderComponent(resolver, defaultSimpleModalOptions) {
        this.resolver = resolver;
        this.defaultSimpleModalOptions = defaultSimpleModalOptions;
        this.modals = [];
        this.previousActiveElement = null;
    }
    SimpleModalHolderComponent.prototype.addModal = function (component, data, options) {
        var _this = this;
        if (!this.viewContainer) {
            return rxjs_1.of(null);
        }
        var factory = this.resolver.resolveComponentFactory(simple_modal_wrapper_component_1.SimpleModalWrapperComponent);
        var componentRef = this.viewContainer.createComponent(factory);
        var modalWrapper = (componentRef.instance);
        var _component = modalWrapper.addComponent(component);
        _component.options = options = Object.assign({}, this.defaultSimpleModalOptions, options);
        modalWrapper.modalClasses = options.wrapperDefaultClasses;
        this.modals.push(_component);
        this.wait().then(function () {
            _this.toggleWrapperClass(modalWrapper.wrapper, options.wrapperClass);
            _this.toggleBodyClass(options.bodyClass);
            _this.wait(options.animationDuration).then(function () {
                _this.autoFocusFirstElement(_component.wrapper, options.autoFocus);
            });
        });
        _component.onClosing(function (modal) { return _this.removeModal(modal); });
        this.configureCloseOnClickOutside(modalWrapper);
        _component.mapDataObject(data);
        return _component.setupObserver();
    };
    SimpleModalHolderComponent.prototype.removeModal = function (closingModal) {
        var _this = this;
        var options = closingModal.options;
        this.toggleWrapperClass(closingModal.wrapper, options.wrapperClass);
        return this.wait(options.animationDuration).then(function () {
            _this.removeModalFromArray(closingModal);
            _this.toggleBodyClass(options.bodyClass);
            _this.restorePreviousFocus();
        });
    };
    SimpleModalHolderComponent.prototype.removeAllModals = function () {
        var _this = this;
        return Promise.all(this.modals.map(function (modal) { return _this.removeModal(modal); }));
    };
    SimpleModalHolderComponent.prototype.toggleBodyClass = function (bodyClass) {
        var _a, _b;
        if (!bodyClass) {
            return;
        }
        var body = document.getElementsByTagName('body')[0];
        var bodyClassItems = bodyClass.split(' ');
        if (!this.modals.length) {
            (_a = body.classList).remove.apply(_a, bodyClassItems);
        }
        else {
            (_b = body.classList).add.apply(_b, bodyClassItems);
        }
    };
    SimpleModalHolderComponent.prototype.configureCloseOnClickOutside = function (modalWrapper) {
        if (modalWrapper.content.options.closeOnClickOutside) {
            modalWrapper.onClickOutsideModalContent(modalWrapper.content.options.closeOnClickOutside, function () {
                modalWrapper.content.close();
            });
        }
    };
    SimpleModalHolderComponent.prototype.autoFocusFirstElement = function (componentWrapper, autoFocus) {
        if (autoFocus) {
            var focusable = componentWrapper.nativeElement.querySelectorAll('button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])');
            if (focusable && focusable.length) {
                this.previousActiveElement = document.activeElement;
                focusable[0].focus();
            }
        }
    };
    SimpleModalHolderComponent.prototype.restorePreviousFocus = function () {
        if (this.previousActiveElement) {
            this.previousActiveElement.focus();
            this.previousActiveElement = null;
        }
    };
    SimpleModalHolderComponent.prototype.toggleWrapperClass = function (modalWrapperEl, wrapperClass) {
        var wrapperClassList = modalWrapperEl.nativeElement.classList;
        var wrapperClassItems = wrapperClass.split(' ');
        if (wrapperClassList.toString().indexOf(wrapperClass) !== -1) {
            wrapperClassList.remove.apply(wrapperClassList, wrapperClassItems);
        }
        else {
            wrapperClassList.add.apply(wrapperClassList, wrapperClassItems);
        }
    };
    SimpleModalHolderComponent.prototype.wait = function (ms) {
        if (ms === void 0) { ms = 0; }
        return new Promise(function (resolve, reject) {
            setTimeout(function () { return resolve(); }, ms);
        });
    };
    SimpleModalHolderComponent.prototype.removeModalFromArray = function (component) {
        var index = this.modals.indexOf(component);
        if (index > -1) {
            this.viewContainer.remove(index);
            this.modals.splice(index, 1);
        }
    };
    __decorate([
        core_1.ViewChild('viewContainer', { read: core_1.ViewContainerRef, static: true }),
        __metadata("design:type", Object)
    ], SimpleModalHolderComponent.prototype, "viewContainer", void 0);
    SimpleModalHolderComponent = __decorate([
        core_1.Component({
            selector: 'simple-modal-holder',
            template: '<ng-template #viewContainer></ng-template>',
        }),
        __param(1, core_1.Inject(simple_modal_options_1.DefaultSimpleModalOptionConfig)),
        __metadata("design:paramtypes", [core_1.ComponentFactoryResolver,
            simple_modal_options_1.SimpleModalOptions])
    ], SimpleModalHolderComponent);
    return SimpleModalHolderComponent;
}());
exports.SimpleModalHolderComponent = SimpleModalHolderComponent;
