// ngx-simple-modal/dist/simple-modal/simple-modal.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("@angular/common");
var core_1 = require("@angular/core");
var simple_modal_holder_component_1 = require("./simple-modal-holder.component");
var simple_modal_wrapper_component_1 = require("./simple-modal-wrapper.component");
var simple_modal_service_1 = require("./simple-modal.service");
var simple_modal_service_factory_1 = require("./simple-modal-service.factory");
var simple_modal_options_1 = require("./simple-modal-options");
var ɵ0 = simple_modal_options_1.defaultSimpleModalOptions;
exports.ɵ0 = ɵ0;
var SimpleModalModule = (function () {
    function SimpleModalModule() {
    }
    SimpleModalModule_1 = SimpleModalModule;
    SimpleModalModule.forRoot = function (config, defaultModalOptions) {
        return {
            ngModule: SimpleModalModule_1,
            providers: [
                { provide: simple_modal_service_1.SimpleModalServiceConfig, useValue: config },
                {
                    provide: simple_modal_service_1.SimpleModalService,
                    useFactory: simple_modal_service_factory_1.SimpleModalServiceFactory,
                    deps: [core_1.ComponentFactoryResolver, core_1.ApplicationRef, core_1.Injector, simple_modal_service_1.SimpleModalServiceConfig],
                },
                {
                    provide: simple_modal_options_1.DefaultSimpleModalOptionConfig,
                    useValue: defaultModalOptions || simple_modal_options_1.defaultSimpleModalOptions,
                },
            ],
        };
    };
    var SimpleModalModule_1;
    SimpleModalModule = SimpleModalModule_1 = __decorate([
        core_1.NgModule({
            declarations: [simple_modal_holder_component_1.SimpleModalHolderComponent, simple_modal_wrapper_component_1.SimpleModalWrapperComponent],
            providers: [
                simple_modal_service_1.SimpleModalService,
                {
                    provide: simple_modal_options_1.DefaultSimpleModalOptionConfig,
                    useValue: ɵ0,
                },
            ],
            imports: [common_1.CommonModule],
            entryComponents: [simple_modal_holder_component_1.SimpleModalHolderComponent, simple_modal_wrapper_component_1.SimpleModalWrapperComponent],
        }),
        __metadata("design:paramtypes", [])
    ], SimpleModalModule);
    return SimpleModalModule;
}());
exports.SimpleModalModule = SimpleModalModule;
