// ngx-simple-modal/dist/simple-modal/simple-modal.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var rxjs_1 = require("rxjs");
var SimpleModalComponent = (function () {
    function SimpleModalComponent() {
    }
    SimpleModalComponent.prototype.mapDataObject = function (data) {
        data = data || {};
        var keys = Object.keys(data);
        for (var i = 0, length_1 = keys.length; i < length_1; i++) {
            var key = keys[i];
            this[key] = data[key];
        }
    };
    SimpleModalComponent.prototype.setupObserver = function () {
        var _this = this;
        return rxjs_1.Observable.create(function (observer) {
            _this.observer = observer;
            _this.completeOnDestroy(_this);
            return function () {
                _this.close();
            };
        });
    };
    SimpleModalComponent.prototype.onClosing = function (callback) {
        this.closerCallback = callback;
    };
    SimpleModalComponent.prototype.close = function () {
        return this.closerCallback(this);
    };
    SimpleModalComponent.prototype.onKeydownHandler = function (evt) {
        if (this.options && this.options.closeOnEscape) {
            this.close();
        }
    };
    SimpleModalComponent.prototype.completeOnDestroy = function (component) {
        var _this = this;
        var ngDestroyOriginal = component.ngOnDestroy;
        component.ngOnDestroy = function () {
            if (ngDestroyOriginal) {
                ngDestroyOriginal.apply(component);
            }
            if (_this.observer) {
                _this.observer.next(_this.result);
                _this.observer.complete();
            }
        };
    };
    SimpleModalComponent.prototype.ngOnDestroy = function () {
    };
    __decorate([
        core_1.HostListener('document:keydown.escape', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [KeyboardEvent]),
        __metadata("design:returntype", void 0)
    ], SimpleModalComponent.prototype, "onKeydownHandler", null);
    return SimpleModalComponent;
}());
exports.SimpleModalComponent = SimpleModalComponent;
